src/storage/garbage_collector.cpp

```diff
@@ -86,80 +120,331 @@ uint32_t GarbageCollector::ProcessUnlinkQueue() {
-        if (visited_slots.insert(undo_record.Slot()).second)
-          TruncateVersionChain(table, undo_record.Slot(), oldest_txn);
-        // Regardless of the version chain we will need to reclaim deleted slots and any dangling pointers to varlens.
-        ReclaimSlotIfDeleted(&undo_record);
-        ReclaimBufferIfVarlen(txn, &undo_record);
+        all_unlinked = all_unlinked && ProcessUndoRecord(&undo_record, &active_txns);
       }
-      txns_to_deallocate_.push_front(txn);
-      txns_processed++;
+      if (all_unlinked) {
+        // We unlinked all of the UndoRecords for this txn, so we can add it to the deallocation queue
+        txns_to_deallocate_.push_front(txn);
+        txns_processed++;
+      } else {
+        // We didn't unlink all of the UndoRecords (UnlinkUndoRecord returned false due to a write-write conflict),
+        // requeue txn for next GC run. Unlinked UndoRecords will be skipped on the next time around since we use the
+        // table pointer of an UndoRecord as the internal marker of being unlinked or not
+        requeue.push_front(txn);
+      }
+    }
+  }
+
+  storage::UndoBuffer *buf = nullptr;
+  std::forward_list<storage::UndoBuffer *> buf_requeue;
+
+  // Process every compaction buffer in the buffer unlink queue
+  while (!buffers_to_unlink_.empty()) {
+    buf = buffers_to_unlink_.front();
+    buffers_to_unlink_.pop_front();
+    if (buf->Empty()) {
+      // This compaction buffer is empty so it is safe to delete it
+      delete buf;
     } else {
-      // This is a committed txn that is still visible, requeue for next GC run
-      requeue.push_front(txn);
+      // This is a list of undo records that may or may not be visible to any running txns. Proceed with unlinking them
+      // with an Interval GC approach
+      bool all_unlinked = true;
+      for (auto &undo_record : *buf) {
+        all_unlinked = ProcessUndoRecord(&undo_record, &active_txns) && all_unlinked;
+      }
+      if (all_unlinked) {
+        // We unlinked all of the UndoRecords for this compaction buffer, so we can add it to the deallocation queue
+        buffers_to_deallocate_.push_front(buf);
+      } else {
+        // We didn't unlink all of the UndoRecords (UnlinkUndoRecord returned false due to a write-write conflict),
+        // requeue buf for next GC run. Unlinked UndoRecords will be skipped on the next time around since we use the
+        // table pointer of an UndoRecord as the internal marker of being unlinked or not
+        buf_requeue.push_front(buf);
+      }
     }
   }
 
-  // Requeue any txns that we were still visible to running transactions
-  txns_to_unlink_ = transaction::TransactionQueue(std::move(requeue));
+  // Requeue any txns that have an undo record still visible to some running transaction
+  if (!requeue.empty()) {
+    txns_to_unlink_ = transaction::TransactionQueue(std::move(requeue));
+  }
```

Comment:
The code flow could be more coherent if you move the above four lines to line 142.